src/dvc_data/build.py

```diff
@@ -77,10 +77,26 @@ def _stage_file(path, fs, name, odb=None, upload_odb=None, dry_run=False):
 
 
 def _build_tree(
-    path, fs, name, ignore: "Ignore" = None, no_progress_bar=False, **kwargs
+    path,
+    fs,
+    fs_info,
+    name,
+    odb=None,
+    ignore: "Ignore" = None,
+    no_progress_bar=False,
+    **kwargs,
 ):
+    from .hashfile.hash_info import HashInfo
     from .objects.tree import Tree
 
+    value = fs_info.get(name)
+    if odb and value:
+        try:
+            tree = Tree.load(odb, HashInfo(name, value))
+            return Meta(nfiles=len(tree)), tree
+        except FileNotFoundError:
+            pass
+
```

Comment:
This temporarily worsens the code by making this function crowded, but this will go away soon.